subsys/dfu/dfu_target/src/dfu_target_smp.c

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <pm_config.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/dfu/mcuboot.h>
+#include <zephyr/mgmt/mcumgr/mgmt/mgmt.h>
+#include <zephyr/mgmt/mcumgr/smp/smp_client.h>
+#include <zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h>
+#include <zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt.h>
+#include <zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt_client.h>
+#include <dfu/dfu_target.h>
+#include <dfu/dfu_target_smp.h>
+#include <zephyr/devicetree.h>
+
+LOG_MODULE_REGISTER(dfu_target_smp, CONFIG_DFU_TARGET_LOG_LEVEL);
+
+#define MCUBOOT_HEADER_MAGIC 0x96f3b83d
+
+struct upload_progress {
+	size_t image_size;
+	size_t offset;
+	uint32_t image_num;
+	int smp_status;
+	int smp_echo;
+};
+
+static struct upload_progress upload_state;
+static struct mcumgr_image_data image_info[2];
+static int image_count;
+static bool recovery_mode_active;
+static dfu_target_reset_cb_t dfu_target_reset_cb;
+static struct smp_client_object smp_client;
+static struct img_mgmt_client img_gr_client;
+static struct os_mgmt_client os_gr_client;
+
+static const char dfu_smp_echo_str[] = "Recovery";
+
+static int dfu_target_smp_recovery_enable(void);
+
+static int img_state_res_update(struct mcumgr_image_state *res_buf)
+{
+	if (res_buf->status) {
+		LOG_ERR("Image state response command fail, err:%d", res_buf->status);
+		image_count = 0;
+	} else {
+		image_count = res_buf->image_list_length;
+	}
+
+	return res_buf->status;
+}
+
+static int img_upload_res_update(struct mcumgr_image_upload *res_buf)
+{
+	if (res_buf->status) {
+		LOG_ERR("Image Upload command fail err:%d", res_buf->status);
+	} else {
+		upload_state.offset = res_buf->image_upload_offset;
+	}
+	return res_buf->status;
+}
+
+int dfu_target_smp_client_init(void)
+{
+	int rc;
+
+	rc = smp_client_object_init(&smp_client, SMP_SERIAL_TRANSPORT);
+	if (rc) {
+		return rc;
+	}
+
+	img_mgmt_client_init(&img_gr_client, &smp_client, 2, image_info);
+	os_mgmt_client_init(&os_gr_client, &smp_client);
+
+	return rc;
+}
+
+bool dfu_target_smp_identify(const void *const buf)
+{
+	/* MCUBoot headers starts with 4 byte magic word */
+	return *((const uint32_t *)buf) == MCUBOOT_HEADER_MAGIC;
+}
+
+int dfu_target_smp_init(size_t file_size, int img_num, dfu_target_callback_t cb)
+{
+	ARG_UNUSED(cb);
+	if (img_num < 0) {
+		return -ENOENT;
+	}
+
+	if (dfu_target_reset_cb) {
+		upload_state.image_num = 2;
```

Comment:
This is also a bit flimsy, works OK for nrf52840dk in normal state but if someone decides to change MCUboot on that core to be e.g. direct XIP, they would want the opposite slot. Probably OK for initial commit but should have a jira created